cmd/telemetry-agent/main.go

```diff
@@ -0,0 +1,280 @@
+			if !errors.Is(err, context.Canceled) {
+				metricsLogger.Errorw("error during sending telemetry", zap.Error(err))
+			}
+			continue
+		}
+
+		// write sent data to history file
+		historyFile := filepath.Join(c.TelemetryHistoryPath, filepath.Base(pillarM.Filename))
+		l.Infow("writing metrics to history file",
+			zap.String("pillar file", pillarM.Filename),
+			zap.String("history file", historyFile))
+		if err := metrics.WriteMetricsToHistory(historyFile, report); err != nil {
+			l.Errorw("failed to write metrics into history file",
+				zap.String("pillar file", pillarM.Filename),
+				zap.String("history file", historyFile),
+				zap.Error(err))
+			continue
+		}
+
+		// remove original Pillar's metrics file
+		l.Infow("removing metrics file", zap.String("file", pillarM.Filename))
+		if err := os.Remove(pillarM.Filename); err != nil {
+			l.Errorw("failed to remove metrics file",
+				zap.String("file", pillarM.Filename),
+				zap.Error(err))
+			continue
+		}
+	}
+	return nil
+}
+
+func main() {
+	verbose := flag.Bool("verbose", false, "enable verbose logging")
+	flag.Usage = func() {
+		flagSet := flag.CommandLine
+		fmt.Fprintf(os.Stderr, "Usage of %s:\n", os.Args[0])
+		order := []string{"verbose"}
+		for _, name := range order {
+			f := flagSet.Lookup(name)
+			fmt.Fprintf(os.Stderr, "-%s\n\t%s\n", f.Name, f.Usage)
+		}
+		fmt.Fprintf(os.Stderr, "\n")
+		fmt.Fprintf(os.Stderr, `Environment variables:
+
+PERCONA_TELEMETRY_ROOT_PATH - define Percona telemetry root path on local filesystem. Default: /usr/local/percona/telemetry
+PERCONA_TELEMETRY_CHECK_INTERVAL - define time interval in seconds for checking Percona Pillars telemetry. Default: 86400 sec
+PERCONA_TELEMETRY_URL - define Percona Platform URL for sending Pillars telemetry to. Default: https://check.percona.com/v1/telemetry/GenericReport
+PERCONA_TELEMETRY_RESEND_INTERVAL - define wait time in seconds to sleep before retrying request to Percona Platform in case of request failure. Default: 60 sec
+PERCONA_TELEMETRY_HISTORY_KEEP_INTERVAL - define time interval in seconds for keeping old history telemetry files on filesystem. Default: 604800 sec (7d)
+
+`)
+	}
+	flag.Parse()
+
+	logger.SetupGlobal(&logger.GlobalOpts{LogName: "telemetry-agent", LogDevMode: false, LogDebug: *verbose})
+	l := zap.L().Sugar()
+	defer func(l *zap.SugaredLogger) {
+		_ = l.Sync()
+	}(l)
+
+	l.Info("parsing env params")
+	conf := config.InitConfig()
+	l.Infow("values from config:", zap.Any("config", conf))
+
+	// check that <telemetry root>/history dir exists on filesystem
+	if err := createTelemetryDirs(conf); err != nil {
+		l.DPanic(err)
+	}
+
+	pltClient, err := createPerconaPlatformClient(conf)
+	if err != nil {
+		l.DPanic(err)
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	l.Info("Percona Telemetry Agent started")
+	var wg sync.WaitGroup
+	wg.Add(1)
+	utils.SignalRunner(
+		func() {
```

Comment:
For me it looks upside down. In most our projects signal handling happens in separate goroutine and main application logic stays straightforward.